### ISSUE TITLE
`ruff format --check` outputs wrong line line/column numbers on syntax errors

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#24528]
### Summary

```py
...
if asdf:
```
`ruff format` correctly says `2:9`:
```console
$ ruff format foo.py
error: Failed to parse foo.py:2:9: Expected an indented block after `if` statement
```
but adding `--check` causes it to incorrectly output `1:1`:
```console
$ ruff format --check foo.py
invalid-syntax: Expected an indented block after `if` statement
--> foo.py:1:1
```

### Version

0.15.8

**Comment by dylwil3** [comment]
Hm, I'm not able to reproduce this. Moreover, the second snippet looks more like the beginning of the output for `ruff check` not `ruff format --check` (but when I tried that I still got the correct line number). Is there a missing step or configuration in the original example?

```console
❯ cat ex.py
...
if asdf:

❯ uvx ruff@0.15.8 format --isolated ex.py
error: Failed to parse ex.py:2:9: Expected an indented block after `if` statement

❯ uvx ruff@0.15.8 format --isolated --check ex.py
error: Failed to parse ex.py:2:9: Expected an indented block after `if` statement
```

**Comment by DetachHead** [comment]
hmm looks like it only happens in preview mode

```toml
# ruff.toml
preview = true
```

**Comment by dylwil3** [comment]
Thank you, that's what I was missing!

**Comment by ntBre** [comment]
I looked into this a bit today (before I noticed you assigned yourself!), and there are a few problems with our current setup. First of all, we're not attaching a range to the diagnostic here:

https://github.com/astral-sh/ruff/blob/2ad94df0e1056aa4da57c66dd51ea2338811136d/crates/ruff/src/commands/format.rs#L884-L890

However, if we try to attach a range there, we'll immediately run into the empty source code passed to `SourceFileBuilder::new` (we try to render something like `source[range]` where `range` is past the end of the empty `source`).

After that, we still run into the `hide_snippet(true)` part, which ends up marking the annotation as being "file-level" and rendered with a location of `1:1` anyway.

I think we can address that last part by changing our behavior to build a real header here:

https://github.com/astral-sh/ruff/blob/e89f8ef295478d58981d19369d2e66f8fe2104ea/crates/ruff_annotate_snippets/src/renderer/display_list.rs#L1237-L1241

either by passing a real `body` to `format_header` or by computing the header information separately (which seems preferable to avoid rendering the body just to process its lines).

**Comment by dylwil3** [comment]
Thanks @ntBre ! I noticed some of the same things.

A minimal hack is to swoop in and re-read the source so you can provide a `Span`, like this:

```diff
diff --git i/crates/ruff/src/commands/format.rs w/crates/ruff/src/commands/format.rs
index 78cab1a17a..c12bde32fc 100644
--- i/crates/ruff/src/commands/format.rs
+++ w/crates/ruff/src/commands/format.rs
@@ -893,11 +893,27 @@ impl From<&FormatCommandError> for Diagnostic {
             FormatCommandError::Ignore(error) => {
                 Diagnostic::new(DiagnosticId::Io, Severity::Error, error)
             }
-            FormatCommandError::Parse(display_parse_error) => Diagnostic::new(
-                DiagnosticId::InvalidSyntax,
-                Severity::Error,
-                &display_parse_error.error().error,
-            ),
+            FormatCommandError::Parse(display_parse_error) => {
+                if let Some(path) = error.path() {
+                    let file = SourceFileBuilder::new(
+                        path.to_string_lossy(),
+                        std::fs::read_to_string(path).unwrap(),
+                    )
+                    .finish();
+                    let span = Span::from(file).with_range(display_parse_error.error().location);
+                    return Diagnostic::invalid_syntax(
+                        span,
+                        display_parse_error.error().error.clone(),
+                        display_parse_error.error().location,
+                    );
+                } else {
+                    Diagnostic::new(
+                        DiagnosticId::InvalidSyntax,
+                        Severity::Error,
+                        &display_parse_error.error().error,
+                    )
+                }
+            }
             FormatCommandError::Panic(path, panic_error) => {
                 return create_panic_diagnostic(panic_error, path.as_deref());
             }

```

This works but feels a little wonky... I wonder if there is a not so painful way I can pass something like the `Span` along with the `ParseError` all the way from the actual format invocation (where we already have the source code in hand)? I'll have to play around a bit more.

**Comment by MichaReiser** [comment]
I haven't looked too closely at the implementation but what I'm wondering:

* Do we need all these variants?
* Could some of them be simple wrappers around `Diagnostic`? Or why is it that we defer constructing the `Diagnostic`?

**Comment by ntBre** [comment]
I don't quite remember why we deferred the `Diagnostic` conversion. It may just be to support the stable behavior of not converting them, but that sounds like a better approach in the long term.

**Comment by dylwil3** [comment]
I think we should fix this at the same time that we stabilize the `ruff format --check` outputs (as in https://github.com/astral-sh/ruff/issues/20755). Once we don't have to support the current stable behavior, we can create the `Diagnostic` much earlier, which will automatically handle the span etc. (We can also remove the struct `DisplayParseError` entirely.)

That does mean it will have to wait for the next minor release, but that's not too far off.

**Comment by ntBre** [comment]
I'm going to consider the immediate issue here closed by https://github.com/astral-sh/ruff/pull/27045 but open a follow-up issue to track using `Diagnostic`s in a better way, which I still think is the better solution.
